aim/web/middlewares/profiler/profiler.py

```diff
@@ -0,0 +1,81 @@
+import os.path
+import time
+import codecs
+from typing import Optional
+from logging import getLogger
+from uuid import uuid4
+
+from pyinstrument import Profiler
+from pyinstrument.renderers import HTMLRenderer
+
+from starlette.routing import Router
+from starlette.requests import Request
+from starlette.types import ASGIApp, Message, Receive, Scope, Send
+
+logger = getLogger("profiler")
+
+
+class PyInstrumentProfilerMiddleware:
+    level = 0
+
+    def __init__(
+            self, app: ASGIApp,
+            *,
+            server_app: Optional[Router] = None,
+            profiler_interval: float = 0.0001,
+            repo_path=None,
+            **profiler_kwargs
+    ):
+        self.app = app
+        self._profiler = Profiler(interval=profiler_interval)
+
+        self._server_app = server_app
+        self._profiler_kwargs: dict = profiler_kwargs
+
+        self._profiler_timestamp = str(time.time())
+        self._profiler_log_path = os.path.join(repo_path, "profiler", self._profiler_timestamp)
+        os.makedirs(self._profiler_log_path, exist_ok=True)
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if scope["type"] != "http":
+            await self.app(scope, receive, send)
+            return
+
+        if self._profiler.is_running:
+            should_stop = False
+        else:
+            should_stop = True
+            self._profiler.start()
+
+        request = Request(scope, receive=receive)
+        method = request.method
+        path = request.url.path
+
+        file_name = (
+            '{timestamp}_{method}_{path}.html'.format(
+                timestamp=time.time(),
+                method=method.lower(),
+                path='_'.join(path.strip('/').split('/')).lower()
+            )
+        )
+
+        # Default status code used when the application does not return a valid response
+        # or an unhandled exception occurs.
+        status_code = 500
+
+        async def wrapped_send(message: Message) -> None:
+            if message['type'] == 'http.response.start':
+                nonlocal status_code
+                status_code = message['status']
+            await send(message)
+
+        try:
+            await self.app(scope, receive, wrapped_send)
+        finally:
+            if scope["type"] == "http" and should_stop:
+                self._profiler.stop()
```

Comment:
I think we can add a line covering request params and body.
Just need to keep HTML syntax.

Reply:
or we could just log it into a separate file with "almost" the same naming. Injecting into html is a bit challenging since its mostly JS code